typing/typecore.ml

```diff
@@ -6194,112 +6231,284 @@ and type_andops env sarg sands expected_ty =
+       "spine-locally", and handle [x]'s mode separately.)  We'll refer to this
+       as the "input mode" below.
+
+     - By the same token, the modes of [body] and the entire comprehension must
+       be *the same as each other*, as we are generating a sequence made up of
+       the result of evaluating [body] repeatedly.  We'll refer to this as the
+       "output mode" below.
+
+     - The input mode must be *below* the output mode.  Clearly, the two can be
+       the same as each other; and if the input is local, then the output surely
+       cannot be global, as it can refer to the input and we cannot have heap
+       pointers pointing into the stack.  However, if the input is global, then
+       it is perfectly safe for a local sequence to contain references to it,
+       and so there is no harm in allowing the output to be local.
+
+     Thus, the question turns on what mode we are to use for the output, the
+     mode of [body] and the entire comprehension.  While it would be nice to be
+     polymorphic here, *we are unfortunately currently constrained to check
+     comprehensions at global mode*.  The reasons for this are separate for list
+     and array comprehensions:
+
+     - For list comprehensions: List comprehensions are desugared in terms of
+       functions from [CamlinternalComprehension]; as regular OCaml functions,
+       they cannot cannot have the desired (or any) mode polymorphism.  The
+       input can either be marked as [local_] (in which case we cannot create a
+       global list with a list comprehension) or not (in which case we cannot
+       use list comprehensions to iterate over local lists).  Similarly, the
+       output can either be marked as [local_] or not, and that fixes where the
+       allocation is to happen.
+
+     - For array comprehensions: We only have global arrays, and do not
+       currently allow there to be such a thing as a local array at all.
+
+     This fully determines the mode situation.  Thus, until we loosen either of
+     these restrictions, we do not pass modes to other functions for
+     typechecking comprehensions.
+
+     In order to understand the reasoning about modes for arrays, anywhere we
+     need to provide modes while typechecking comprehensions, we will reference
+     this comment by its incipit (the initial question, right at the start). *)
+  let new_env, comp_clauses =
+    type_comprehension_clauses
+      ~loc ~env ~comprehension_type ~container_type clauses
+  in
+  let comp_body =
+    (* To understand why comprehension bodies are checked at [mode_global], see
+       "What modes should comprehensions use?", above *)
+    type_expect new_env mode_global sbody (mk_expected element_ty)
+  in
+  re { exp_desc       = make_texp { comp_body ; comp_clauses }
+     ; exp_loc        = loc
+     ; exp_extra      = []
+     ; exp_type       = instance (container_type comp_body.exp_type)
```

Comment:
That's because `Pexp_tuple` unifies generic types, which Leo describes as something of an anti-pattern. I *do* think we need it here, which does not unify generic types.